docs/design/api-plugin-model.md

```diff
@@ -0,0 +1,139 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/design/api-plugin-model.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+# Kubernetes API Plugin Model
+
+## Goals
+
+* Allow new APIs to be registered and serviced by third party components
+  dynamically at run time, not requiring a recompilation of any core kubernetes
+  component.
+* All state is stored together by apiserver; third party components can remain
+  stateless.
+* Allow core components to be examined by third parties via a well defined API.
+
+## New REST Objects
+
+1. The `Provider` type. The provider specifies:
+  1. A REST endpoint to dial out to.
+  2. A list of types for which this provider wants web hooks. 'all' is a valid
+     choice.
+  3. Optionally, for each type, a list of hooks the provider wants called.
+2. The `ThirdPartyType` type. This declares a new type to the system, and stores
+   all the information needed to identify this type.
+  1. APIGroup
+  2. Kind
+  3. The version of this type.
+  4. The above must specify a unique object (maybe we should pack it into the
+     name, as well as having it in separate fields for convenience).
+  5. Whether or not the resource is namespaced (ThirdPartyType objects are *not*
+     themselves in a namespace, like Node and Namespace objects.)
+  6. Swagger spec for this type (or a `Provider` from which to get this spec).
+     Alternatively, some other method of specification. But apiserver needs to
+     be able to integrate this information into the swagger docs it exports.
+  7. A choice: validation could be performed by a hook (see below), or by
```

Comment:
There are multiple kinds of validation:
1. Schema validation. We do this only in a rudimentary way -- whatever the standard JSON parser barfs on, such as missing required fields.
2. Value validation: More specific value rules (e.g., DNS labels) and multi-field relationships.

I would like to at least optionally perform more strict schema validation and would like to use more of the swagger spec to drive some of the value validation. However, ideally this would not work in a way that's substantively different than what we do for built-in resources.